src/soc/ibm/power9/homer.c

```diff
@@ -997,6 +1090,1376 @@ static void istep_16_1(int this_core)
+			 * here:
+			 * 0      : PPM Write control
+			 * 9      : FUSED_CORE_MODE
+			 * 10     : STOP_EXIT_TYPE_SEL
+			 * 13     : WKUP_NOTIFY_SELECT
+			 */
+
+			/* Clear Core PPM Mode register */
+			write_scom_for_chiplet(core_chiplet, C_CPPM_CPMMR_CLEAR, 0x401b000000000000);
+
+			/* Clear Core PPM Errors */
+			write_scom_for_chiplet(core_chiplet, C_CPPM_ERR, 0);
+
+			/*
+			 * Clear Hcode Error Injection and other CSAR settings:
+			 *  - CPPM_CSAR_FIT_HCODE_ERROR_INJECT
+			 *  - CPPM_CSAR_ENABLE_PSTATE_REGISTRATION_INTERLOCK
+			 *  - CPPM_CSAR_PSTATE_HCODE_ERROR_INJECT
+			 *  - CPPM_CSAR_STOP_HCODE_ERROR_INJECT
+			 * CPPM_CSAR_DISABLE_CME_NACK_ON_PROLONGED_DROOP is NOT cleared
+			 * as this is a persistent, characterization setting.
+			 */
+			write_scom_for_chiplet(core_chiplet, C_CPPM_CSAR_CLEAR, 0x1b00000000);
+
+			/* Restore CORE PPM Error Mask */
+			err_mask = 0xfff00000; // from Hostboot's log
+			write_scom_for_chiplet(core_chiplet, C_CPPM_ERRMSK, (uint64_t)err_mask << 32);
+		}
+	}
+}
+
+static void pstate_gpe_init(struct homer_st *homer, uint64_t cores)
+{
+	enum {
+		/* The following constants hold approximate values */
+		PGPE_TIMEOUT_MS  = 500,
+		PGPE_POLLTIME_MS = 20,
+		TIMEOUT_COUNT    = PGPE_TIMEOUT_MS / PGPE_POLLTIME_MS,
+
+		EQ_QPPM_QPMMR = 0x100F0103,
+
+		PU_GPE2_PPE_XIXCR    = 0x00064010,
+		PU_GPE2_PPE_XIDBGPRO = 0x00064015,
+		PU_GPE3_PPE_XIDBGPRO = 0x00066015,
+
+		PU_GPE2_GPEIVPR_SCOM    = 0x00064001,
+		PU_OCB_OCI_OCCS2_SCOM   = 0x0006C088,
+		PU_OCB_OCI_OCCFLG_SCOM2 = 0x0006C08C,
+		PU_GPE2_GPETSEL_SCOM    = 0x00064000,
+
+		/* OCC SCRATCH2 */
+		PGPE_ACTIVE                 = 0,
+		PGPE_PSTATE_PROTOCOL_ACTIVE = 1,
+
+		/* XSR */
+		HALTED_STATE = 0,
+
+		/* XCR */
+		RESUME         = 2,
+		TOGGLE_XSR_TRH = 4,
+		HARD_RESET     = 6,
+	};
+
+	uint64_t occ_scratch;
+	/* ATTR_VDD_AVSBUS_BUSNUM */
+	uint8_t avsbus_number = 0;
+	/* ATTR_VDD_AVSBUS_RAIL */
+	uint8_t avsbus_rail = 0;
+
+	uint64_t ivpr = 0x80000000 + offsetof(struct homer_st, ppmr.l1_bootloader);
+	write_scom(PU_GPE2_GPEIVPR_SCOM, ivpr << 32);
+
+	/* Set up the OCC Scratch 2 register before PGPE boot */
+	occ_scratch = read_scom(PU_OCB_OCI_OCCS2_SCOM);
+	occ_scratch &= ~PPC_BIT(PGPE_ACTIVE);
+	occ_scratch &= ~PPC_BITMASK(27, 32);
+	occ_scratch |= PPC_PLACE(avsbus_number, 27, 1);
+	occ_scratch |= PPC_PLACE(avsbus_rail, 28, 4);
+	write_scom(PU_OCB_OCI_OCCS2_SCOM, occ_scratch);
+
+	write_scom(PU_GPE2_GPETSEL_SCOM, 0x1A00000000000000);
+
+	/* OCCFLG2_PGPE_HCODE_FIT_ERR_INJ | OCCFLG2_PGPE_HCODE_PSTATE_REQ_ERR_INJ */
+	write_scom(PU_OCB_OCI_OCCFLG2_CLEAR, 0x1100000000);
+
+	printk(BIOS_ERR, "Attempting PGPE activation...\n");
+
+	write_scom(PU_GPE2_PPE_XIXCR, PPC_PLACE(HARD_RESET, 1, 3));
+	write_scom(PU_GPE2_PPE_XIXCR, PPC_PLACE(TOGGLE_XSR_TRH, 1, 3));
+	write_scom(PU_GPE2_PPE_XIXCR, PPC_PLACE(RESUME, 1, 3));
+
+	/* Does this need to be such a long time? */
+	wait_ms(20*TIMEOUT_COUNT,
```

Comment:
```suggestion
	wait_ms(PGPE_POLLTIME_MS * TIMEOUT_COUNT,
```

I wonder if we should add a version of wait_ms() to coreboot that checks condition periodically instead of using busy loop. For memory training busy reading significantly slowed down the training. The issue is that we can't use delay() in ramstage, at least not in the same form. Still, this is not part of this task, I'm just thinking out loud :slightly_smiling_face: